### PR TITLE
RFC: codec: cadence: Add codec API functions symbols

### PR DESCRIPTION
**Pull request by dbaluta**
Cadence library comes with various codecs. In order to support one
codec we need to initialize the API member of cadence_api with exported
library function.

This patch adds Cadence exported functions for aac, bsac, dabplus, drm
and mp3 codecs.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

### CODE REVIEW

**Comment by dbaluta** [comment]
@mrajwa does your cadence library really export a function named `cadence_api_function` or is just a wrapper over cadence stuff?

**Comment by mrajwa** [comment]
> @mrajwa does your cadence library really exports a function named `cadence_api_function` or is just a wrapper over cadence stuff?

It is just a placeholder for the exact name. It was written in this way because the API may be restricted. That said are you sure you can publish names like `xa_aac_dec`?

File: src/audio/codec_adapter/codec/cadence.c [hunk 1]
```diff
@@ -20,7 +20,27 @@
 static struct cadence_api cadence_api_table[] = {
 	{
 		.id = 0x01,
-		.api = cadence_api_function
+		.api = xa_aac_dec,
+	},
+	{
+		.id = 0x02,
```
Comment (mrajwa) [inline src/audio/codec_adapter/codec/cadence.c:26]: I assume all these are in fact different cadence codecs performing different processing?

**Review by mrajwa — APPROVED** [review]

File: src/audio/codec_adapter/codec/cadence.c (hunk 1 above)
Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:26]: yes.

Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:26]: @mrajwa one more thing. I think we can still keep cadence_api_function symbol as a wildcard.

Comment (mrajwa) [reply src/audio/codec_adapter/codec/cadence.c:26]: @dbaluta yes good idea, please don't change the 0x01 ID and keep is as a wildcard.

**Comment by lgirdwood** [comment]
@cujomalainey fyi

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@dbaluta CI build shows
```
[100%] Linking C executable sof
/usr/local/bin/cmake -E cmake_link_script CMakeFiles/sof.dir/link.txt --verbose=1
/localdisk/tools/xtensa/RG-2017.8-linux/XtensaTools/bin/xt-xcc    CMakeFiles/sof.dir/src/platform/cannonlake/lib/clk.c.o CMakeFiles/sof.dir/src/platform/intel/cavs/lib/clk.c.o CMakeFiles/sof.dir/src/platform/intel/cavs/lib/dai.c.o CMakeFiles/sof.dir/src/platform/intel/cavs/lib/dma.c.o CMakeFiles/sof.dir/src/platform/intel/cavs/lib/memory.c.o CMakeFiles/sof.dir/src/platform/intel/cavs/lib/pm_runtime.c.o CMakeFiles/sof.dir/src/platform/intel/cavs/lib/pm_memory.c.o CMakeFiles/sof.dir/src/platform/intel/cavs/lib/mem_window.c.o CMakeFiles/sof.dir/src/platform/intel/cavs/platform.c.o CMakeFiles/sof.dir/src/arch/xtensa/drivers/interrupt.c.o CMakeFiles/sof.dir/src/arch/xtensa/drivers/timer.c.o CMakeFiles/sof.dir/src/arch/xtensa/lib/notifier.c.o CMakeFiles/sof.dir/src/arch/xtensa/lib/cpu.c.o CMakeFiles/sof.dir/src/arch/xtensa/schedule/schedule.c.o CMakeFiles/sof.dir/src/arch/xtensa/schedule/task.c.o CMakeFiles/sof.dir/src/arch/xtensa/xtos/crt1-boards.S.o CMakeFiles/sof.dir/src/arch/xtensa/xtos/_vectors.S.o CMakeFiles/sof.dir/src/arch/xtensa/init.c.o CMakeFiles/sof.dir/src/arch/xtensa/exc-dump.S.o CMakeFiles/sof.dir/src/arch/xtensa/main-entry.S.o CMakeFiles/sof.dir/src/ipc/ipc.c.o CMakeFiles/sof.dir/src/ipc/handler.c.o CMakeFiles/sof.dir/src/ipc/dma-copy.c.o CMakeFiles/sof.dir/src/audio/host.c.o CMakeFiles/sof.dir/src/audio/pipeline.c.o CMakeFiles/sof.dir/src/audio/pipeline_static.c.o CMakeFiles/sof.dir/src/audio/component.c.o CMakeFiles/sof.dir/src/audio/buffer.c.o CMakeFiles/sof.dir/src/audio/channel_map.c.o CMakeFiles/sof.dir/src/audio/volume/volume_generic.c.o CMakeFiles/sof.dir/src/audio/volume/volume_hifi3.c.o CMakeFiles/sof.dir/src/audio/volume/volume.c.o CMakeFiles/sof.dir/src/audio/src/src_generic.c.o CMakeFiles/sof.dir/src/audio/src/src_hifi2ep.c.o CMakeFiles/sof.dir/src/audio/src/src_hifi3.c.o CMakeFiles/sof.dir/src/audio/src/src.c.o CMakeFiles/sof.dir/src/audio/eq_fir/eq_fir.c.o CMakeFiles/sof.dir/src/audio/eq_fir/eq_fir_generic.c.o CMakeFiles/sof.dir/src/audio/eq_fir/eq_fir_hifi2ep.c.o CMakeFiles/sof.dir/src/audio/eq_fir/eq_fir_hifi3.c.o CMakeFiles/sof.dir/src/audio/eq_iir/eq_iir.c.o CMakeFiles/sof.dir/src/audio/eq_iir/iir.c.o CMakeFiles/sof.dir/src/audio/dcblock/dcblock.c.o CMakeFiles/sof.dir/src/audio/dcblock/dcblock_generic.c.o CMakeFiles/sof.dir/src/audio/tdfb/tdfb.c.o CMakeFiles/sof.dir/src/audio/tdfb/tdfb_generic.c.o CMakeFiles/sof.dir/src/audio/tdfb/tdfb_hifiep.c.o CMakeFiles/sof.dir/src/audio/tdfb/tdfb_hifi3.c.o CMakeFiles/sof.dir/src/audio/tone.c.o CMakeFiles/sof.dir/src/audio/mixer.c.o CMakeFiles/sof.dir/src/audio/mux/mux.c.o CMakeFiles/sof.dir/src/audio/mux/mux_generic.c.o CMakeFiles/sof.dir/src/audio/switch.c.o CMakeFiles/sof.dir/src/audio/dai.c.o CMakeFiles/sof.dir/src/audio/kpb.c.o CMakeFiles/sof.dir/src/audio/selector/selector_generic.c.o CMakeFiles/sof.dir/src/audio/selector/selector.c.o CMakeFiles/sof.dir/src/audio/pcm_converter/pcm_converter.c.o CMakeFiles/sof.dir/src/audio/pcm_converter/pcm_converter_generic.c.o CMakeFiles/sof.dir/src/audio/pcm_converter/pcm_converter_hifi3.c.o CMakeFiles/sof.dir/src/audio/asrc/asrc.c.o CMakeFiles/sof.dir/src/audio/asrc/asrc_farrow.c.o CMakeFiles/sof.dir/src/audio/asrc/asrc_farrow_generic.c.o CMakeFiles/sof.dir/src/audio/asrc/asrc_farrow_hifi3.c.o CMakeFiles/sof.dir/src/audio/codec_adapter/codec_adapter.c.o CMakeFiles/sof.dir/src/audio/codec_adapter/codec/generic.c.o CMakeFiles/sof.dir/src/audio/codec_adapter/codec/cadence.c.o CMakeFiles/sof.dir/src/lib/agent.c.o CMakeFiles/sof.dir/src/lib/lib.c.o CMakeFiles/sof.dir/src/lib/alloc.c.o CMakeFiles/sof.dir/src/lib/notifier.c.o CMakeFiles/sof.dir/src/lib/pm_runtime.c.o CMakeFiles/sof.dir/src/lib/clk.c.o CMakeFiles/sof.dir/src/lib/dma.c.o CMakeFiles/sof.dir/src/lib/dai.c.o CMakeFiles/sof.dir/src/lib/wait.c.o CMakeFiles/sof.dir/src/math/numbers.c.o CMakeFiles/sof.dir/src/math/trig.c.o CMakeFiles/sof.dir/src/math/decibels.c.o CMakeFiles/sof.dir/src/math/iir_df2t_generic.c.o CMakeFiles/sof.dir/src/math/iir_df2t_hifi3.c.o CMakeFiles/sof.dir/src/math/fir_generic.c.o CMakeFiles/sof.dir/src/math/fir_hifi2ep.c.o CMakeFiles/sof.dir/src/math/fir_hifi3.c.o CMakeFiles/sof.dir/src/samples/audio/smart_amp_test.c.o CMakeFiles/sof.dir/src/samples/audio/detect_test.c.o CMakeFiles/sof.dir/src/spinlock.c.o CMakeFiles/sof.dir/src/drivers/intel/cavs/interrupt.c.o CMakeFiles/sof.dir/src/drivers/intel/cavs/timer.c.o CMakeFiles/sof.dir/src/drivers/intel/cavs/timestamp.c.o CMakeFiles/sof.dir/src/drivers/intel/cavs/ipc.c.o CMakeFiles/sof.dir/src/drivers/intel/cavs/idc.c.o CMakeFiles/sof.dir/src/drivers/intel/hda/hda-dma.c.o CMakeFiles/sof.dir/src/drivers/intel/hda/hda.c.o CMakeFiles/sof.dir/src/drivers/intel/ssp/ssp.c.o CMakeFiles/sof.dir/src/drivers/intel/ssp/mn.c.o CMakeFiles/sof.dir/src/drivers/intel/alh.c.o CMakeFiles/sof.dir/src/drivers/intel/dmic.c.o CMakeFiles/sof.dir/src/drivers/dw/dma.c.o CMakeFiles/sof.dir/src/drivers/interrupt.c.o CMakeFiles/sof.dir/src/trace/dma-trace.c.o CMakeFiles/sof.dir/src/trace/trace.c.o CMakeFiles/sof.dir/src/debug/panic.c.o CMakeFiles/sof.dir/src/init/init.c.o CMakeFiles/sof.dir/src/schedule/dma_single_chan_domain.c.o CMakeFiles/sof.dir/src/schedule/edf_schedule.c.o CMakeFiles/sof.dir/src/schedule/ll_schedule.c.o CMakeFiles/sof.dir/src/schedule/schedule.c.o CMakeFiles/sof.dir/src/schedule/task.c.o CMakeFiles/sof.dir/src/schedule/timer_domain.c.o CMakeFiles/sof.dir/src/idc/idc.c.o  -o sof -Wl,-Map=sof.map -T/quickbuild/workspace1/24733/SOF_FW/build_fw/cannonlake.x src/platform/intel/cavs/lib/libpdown.a src/platform/intel/cavs/libcavs_ext_manifest.a src/arch/xtensa/xtos/libxtos.a src/arch/xtensa/hal/libhal.a src/arch/xtensa/xtos/libxlevel2.a src/arch/xtensa/xtos/libxlevel5.a src/init/libext_manifest.a -nostdlib -lgcc -Wl,--no-check-sections -ucall_user_start -Wl,-static 
CMakeFiles/sof.dir/src/audio/codec_adapter/codec/cadence.c.o:(.data+0x4): undefined reference to `xa_aac_dec'
CMakeFiles/sof.dir/src/audio/codec_adapter/codec/cadence.c.o:(.data+0xc): undefined reference to `xa_bsac_dec'
CMakeFiles/sof.dir/src/audio/codec_adapter/codec/cadence.c.o:(.data+0x14): undefined reference to `xa_dabplus_dec'
CMakeFiles/sof.dir/src/audio/codec_adapter/codec/cadence.c.o:(.data+0x1c): undefined reference to `xa_drm_dec'
CMakeFiles/sof.dir/src/audio/codec_adapter/codec/cadence.c.o:(.data+0x24): undefined reference to `xa_mp3_dec'
CMakeFiles/sof.dir/src/audio/codec_adapter/codec/cadence.c.o:(.data+0x2c): undefined reference to `xa_sbc_dec'
CMakeFiles/sof.dir/build.make:1709: recipe for target 'sof' failed
make[3]: *** [sof] Error 2
```
Looks like we may need a Kconfig to build this in alongside the Cadence binaries ?

**Comment by cujomalainey** [comment]
@chiang831 FYI for potential decoding

**Comment by dbaluta** [comment]
@lgirdwood fixed now.

File: src/audio/codec_adapter/codec/cadence.c [hunk 2]
```diff
@@ -22,6 +22,42 @@ static struct cadence_api cadence_api_table[] = {
 		.id = 0x01,
 		.api = cadence_api_function
 	},
+#ifdef CONFIG_CADENCE_CODEC_AAC_DEC
```
Comment (dbaluta) [inline src/audio/codec_adapter/codec/cadence.c:25]: @mrajwa we might also want to add an ``ifdef`` for ``cadence_api_function``, I couldn't find a proper name for it. But we can do it in a separate patch.

**Comment by dbaluta** [comment]
SOFCI TEST

File: src/audio/codec_adapter/codec/cadence.c (hunk 2 above)
Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:27]: @mrajwa  later edit, added  a separate commit for `cadence_api_function` in this way we avoid compilation errors.

**Comment by lgirdwood** [comment]
@zrombel we are getting an unrelated failure on JSL. Good to merge ?

**Comment by dbaluta** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
Jenkins CI alsabat known and fix pending. @zrombel CI has been pending a while now ?

**Comment by dbaluta** [comment]
@lgirdwood  can we go with this change. It will not impact our future refactor AFAICT. The config symbols will still be needed to select support for a specific codec and the multiplexing codec we talked about will still need to use some ids for codec types.

**Comment by mrajwa** [comment]
@dbaluta no, tests are failing -> cadence.c:98   ERROR cadence_codec_init(): could not find API function for id 1

**Comment by dbaluta** [comment]
@mrajwa is this your internal test right? Do you have CADENCE_CODEC_WRAPPER set to y?

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
> @lgirdwood can we go with this change. It will not impact our future refactor AFAICT. The config symbols will still be needed to select support for a specific codec and the multiplexing codec we talked about will still need to use some ids for codec types.

Ack, I wanted to get the functionality merged so it would be obvious to all developers working on updates on how updates would impact multiple use cases we  have.

**Comment by lgirdwood** [comment]
Just pending on CI.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
Try CI again.

**Comment by lgirdwood** [comment]
Unrelated report on ICL relating to kernel suspend.
